EzPC/codegensecurenn.ml

```
@@ -0,0 +1,724 @@
+
+  | Const (Int64C n) -> seq (o_str (" (int64_t)")) (o_int64 n)
+
+  | Const (UInt32C n) -> seq (o_str (" (uint32_t)")) (o_uint32 n)
+
+  | Const (UInt64C n) -> seq (o_str (" (uint64_t)")) (o_uint64 n)
+
+  | Const (BoolC b) -> o_bool b
+    
+  | Var s -> o_var s
+
+  | Unop (op, e, Some Public) -> seq (o_punop op) (seq o_space (o_expr e))
+  
+  | Unop (op, e, Some (Secret s)) -> o_sunop s op (o_expr e)
+
+  | Binop (op, e1, e2, Some Public) ->
+     o_paren (match op with
+              | R_shift_l -> o_app (o_str "public_lrshift") [o_expr e1; o_expr e2]
+              | Pow -> o_app (o_str "pow") [o_expr e1; o_expr e2]
+              | _ -> seq (o_expr e1) (seq o_space (seq (o_pbinop op) (seq o_space (o_expr e2)))))
+
+  | Binop (op, e1, e2, Some (Secret s)) -> o_secret_binop g op s e1 e2
+
+  | Conditional (e1, e2, e3, lopt) -> o_codegen_expr (Conditional_codegen (Base_e e1, Base_e e2, Base_e e3, (get_opt lopt)))
+
+  | Array_read (e1, e2) -> seq (o_expr e1) (seq (o_str "[") (seq (o_expr e2) (o_str "]")))
+
+  | App (f, args) -> o_app (o_str f) (List.map o_expr args)
+
+  | Subsumption (e, l1, Secret l2) -> o_subsumption l1 l2 (typeof_expr g e |> get_opt) (o_expr e)
+
+  | _ -> failwith "o_expr: impossible branch"
+
+and o_codegen_expr (g:gamma) (e:codegen_expr) :comp =
+  let o_expr = o_expr g in
+  let o_codegen_expr = o_codegen_expr g in
+  match e with
+  | Base_e e -> o_expr e
+              
+  | Input_g (r, sl, s, bt) -> o_str s.name
+
+  | Dummy_g (sl, bt) -> o_str "0"
+    
+  | Output_g (r, sl, e) -> 
+      let o_reveal_bitmask = 
+        match r with 
+        | Client -> o_str "2" (*Bitmask = 10*)
+        | Server -> o_str "1" (*Bitmask = 01*)
+        | Both -> o_str "3" (*Bitmask = 11*)
+      in
+      o_cbfunction sl (o_str "funcReconstruct2PCCons") [o_codegen_expr e; o_reveal_bitmask]
+
+  | Clear_val (e, bt) -> seq (o_codegen_expr e) (seq (o_str "->get_clear_value<") (seq (o_basetyp bt) (o_str ">()")))

```

If the secureNN backend doesn't use `Clear_val`, error out instead of doing something by default. Same applies to other cases.